### PR TITLE
Adding MessageDelete and MessageUpdate activities.

### PR DESCRIPTION
**Pull request by Jeffders**
Adding MessageDelete and MessageUpdate activities.

### CODE REVIEW

File: src/directLine.ts [hunk 1]
```diff
@@ -202,6 +202,24 @@ export interface Message extends IActivity {
...
+    speak?: string,
+    inputHint?: string,
+    value?: object
+}
+
+export interface MessageDelete extends IActivity {
+    type: "delete"
+}
+
 export interface Typing extends IActivity {
```
Comment (billba) [inline src/directLine.ts]: This looks a whole lot like Message. Any reason not to share (most of) a definition?

Comment (Jeffders) [inline src/directLine.ts]: They are the same, except for the type. I will create an IBaseMessage interface with the common pieces.

Comment (billba) [inline src/directLine.ts]: Perfect.

File: src/directLine.ts [hunk 2]
```diff
@@ -520,6 +533,49 @@ export class DirectLine implements IBotConnection {
...
+                    "Content-Type": "application/json",
+                    "Authorization": `Bearer ${this.token}`
+                }
+            })
+            .map(ajaxResponse => ajaxResponse.response.id as string)
+            .catch(error => this.catchPostError(error))
+        )
+        .catch(error => this.catchExpiredToken(error));
+    }
+
```
Comment (billba) [inline src/directLine.ts]: This looks fine from a code perspective, but it's hard to evaluate without understanding the bigger picture. I don't understand who would call these - adaptive cards, I'm guessing?

Also, is there no corresponding `putMessageWithAttachments` for updating messages with well attachments?

**Comment by avilde** [comment]
- What is the progress of this pull request?
- Will it be implemented?
- 

Allowing to update/delete an activity would greatly improve the control of conversation - disable (or hide) actions on an activity and not show activities which are obsolete (like if we want to edit user response).

@billba,  I have a question (hope you can answer it):
I have tried the [endpoints](https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-connector-api-reference?view=azure-bot-service-4.0#update-activity):
`PUT /v3/conversations/{conversationId}/activities/{activityId}`
`DELETE /v3/conversations/{conversationId}/activities/{activityId}`
`serviceUrl` = https://directline.botframework.com/

Using both Bot Secret/Bot Token with authorization `Bearer Token` I always get response **401 Unauthorized**. What I'm doing wrong?

**Comment by billba** [comment]
This is — obviously — a very old PR, and I’m no longer involved with this repo.

**Comment by cwhitten** [comment]
Closing as we've drifted quite a bit from 2017

**Pull request closed without merging**